unified-runtime/source/common/cuda-hip/stream_queue.hpp

```diff
@@ -21,7 +21,10 @@ using ur_stream_guard = std::unique_lock<std::mutex>;
 /// backend 'stream' objects.
 ///
 /// This class is specifically designed for the CUDA and HIP adapters.
-template <typename ST, int CS, int TS> struct stream_queue_t {
+template <typename ST, int CS, int TS, typename BackendData>
```

Comment:
I actually had started a patch moving `BarrierEvent` and `BarrierEventTmp` in here, just passing the event type as a template paramter. Could you do that instead? The `BackendData` trick is pretty smart, but just passing the backend event type seems simpler.